numpy/linalg/_linalg.py

```diff
@@ -815,8 +815,53 @@ def cholesky(a):
     return wrap(r.astype(result_t, copy=False))
 
 
+# outer product
+
+
+def _outer_dispatcher(x1, x2):
+    return (x1, x2)
+
+
+@array_function_dispatch(_outer_dispatcher)
+def outer(x1, x2, /):
+    """
+    Compute the outer product of two vectors.
+
+    This function is Array API compatible. Compared to ``np.outer``
+    it accepts 1-dimensional inputs only.
+
+    Parameters
+    ----------
+    x1 : (M,) array_like
+        One-dimensional input array of size ``N``.
+        Must have a numeric data type.
+    x2 : (N,) array_like
+        One-dimensional input array of size ``M``.
+        Must have a numeric data type.
+
+    Returns
+    -------
+    out : (M, N) ndarray
+        ``out[i, j] = a[i] * b[j]``
+
+    See also
+    --------
+    outer
+
+    """
+    x1 = asarray(x1)
```

Comment:
@mtsokol - only seeing this now since it entered nightly and hence caused astropy to notice it is missing coverage: is there any reason this is not `asanyarray`? It means we have to cover this function even though we already cover `np.outer`, so with `np.asanyarray` this would just have worked. I think for new code we should just assume subclasses are OK and able to deal with standard numpy code.

Reply:
It indeed seems reasonable to accept subclasses in new functions. However, as long as `np.matrix` and masked arrays are still around, I think it'd be a lot safe to have a new validation function that does what `asanyarray` does plus explicitly rejects instances of those two problematic subclasses. And in order to do so, we need a faster check than `isinstance` calls (perhaps adding a private `._invalid_subclass` and checking for that attribute?). WDYT @mhvk?

Reply:
Let's just ignore those?  Concretely:
1. Matrix is already rejected here and even if it wasn't, I wouldn't want to bother: we are trying to phase it out, don't use it with new functions.
2. Masked arrays are a good example of a place where using `asarray()` is guaranteed to give bad results, while `asanyarray()` at least has a *chance* to do the right thing (and in fact, it looks like it does something pretty reasonable here).

The question is what pattern we want, which is the `_invalid_subclass` thought.  I think the answer is: it's the subclasses problem, _especially_ for new functions.  We just *always* use `asanyarray()`, the subclass should/must implement `__array_function__` and deal with it if that doesn't work out.

Using `asanyarray()` gives the subclass a chance at gambling it can use `super().__array_function__` or call `func._implementation()` (and a bit of a gamble it is, as we may change the implementation; astropy is happy with that gamble).

The main thing I do not quite like about the pattern is that it would be nice if we could open this up more clearly to non-subclasses but...

PS: I actually do wonder if rather than trying to implement a proper `__array_function__` for masked arrays, it may be plausible to implement one that does nothing but reject clearly broken ones.